File: InsertInterval.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    # @param intervals, a list of Intervals
    # @param newInterval, a Interval
    # @return a list of Interval
    def insert(self, intervals, newInterval):
        result = []
        for interval in intervals:
            if interval.end < newInterval.start:
                result.append(interval)
            elif interval.start > newInterval.end:
                result.append(copy.copy(newInterval))
                newInterval.start = interval.start
                newInterval.end = interval.end
            elif interval.start <= newInterval.start or interval.end >= newInterval.end:
                newInterval.start = min(interval.start, newInterval.start)
                newInterval.end = max(interval.end, newInterval.end)
        result.append(newInterval)
        return result

File: InsertInterval.cc
/**
* Definition for an interval.
* struct Interval {
*     int start;
*     int end;
*     Interval() : start(0), end(0) {}
*     Interval(int s, int e) : start(s), end(e) {}
* };
*/
class Solution {
public:
    vector<Interval> insert(vector<Interval> &intervals, Interval newInterval) {
        int index = 0;
        vector<Interval> result;
        while(index < intervals.size()) {
            Interval current_interval = intervals[index];
            if (current_interval.start > newInterval.end) {
                result.push_back(newInterval);
                newInterval.start = current_interval.start;
                newInterval.end = current_interval.end;
                index++;
            } else if (current_interval.end < newInterval.start) {
                result.push_back(current_interval);
                index++;
            } else {
                newInterval.start = min(newInterval.start, current_interval.start);
                newInterval.end = max(newInterval.end, current_interval.end);
                index++;
            }
        }
        result.push_back(newInterval);
        return result;
    }
};
